### DISCUSSION TITLE
Button event launched incorrectly

### DISCUSSION THREAD

**Discussion by RafaRioFal4699** [kivy/kivy-github-discussions#16] [Q&A]
Well, I'm here because the Github Actions bot decided my issue [#8482](https://github.com/kivy/kivy/issues/8482) is not a bug, so I'll try. All the info is in the link, but basically an event declared in a button is not fired only in the button, but also in the entire screen

**Comment by Julian-O** [comment]
> The expected behavior is the on_open_menu event only fired when I drag my click FROM INSIDE the button TO THE OUTSIDE. But it's triggering also when I start the drag OUTSIDE of the button.

I am not an expert in this area, but perhaps you can explain *why* you expect that, because it seems counter-intuitive to me. Why wouldn't you be told that a drag is happening above your widget?

Why aren't you monitoring `on_touch_down` instead?

**Reply by RafaRioFal4699** [reply]
`Why wouldn't you be told that a drag is happening above your widget?`
Yes, that makes sense. But that doesn't answer my question. If I assign an event like on_touch_move to the button, at least should it be fired in first place by the button, right? Is the button who should call that event (and actually it is). What doesn't make sense is why if I drag from X to Y, being both outside the bounds of the button, the event is fired with the button as a sender? I am moving outside the button all the time.

`Why aren't you monitoring on_touch_down instead?`
As I say in the issue, my goal is to get an event which only fires when I move from inside the button to outside (either outside-outside nor inside-inside are valid to this case). **Update:** I didn't get this advice the first time I read it. With this, I found a workaround, but I still don't understand this behavior.

I'm talking from my experience with other frameworks, but maybe I'm wrong, or maybe Kivy manage widgets and events slightly different... In that case, all lessons are welcome 👐

**Reply by Julian-O** [reply]
When you receive an `on_touch_move`, you are calling: `root.on_open_menu(self, args[1], self.pos, self.size)`.

Then you define on_open_menu as:  `def on_open_menu(self, sender, touch, *args)`

So, those parameters are: self=root, sender=\<button instance>, touch=on_touch_move's second argument (which is called touch), and args which is a tuple of \<button instance>.pos, <button instance>.size.

The value of sender is unrelated to the message. It is always the button instance.

**Reply by Julian-O** [reply]
> As I say in the issue, my goal is to get an event which only fires when I move from inside the button to outside

If you want it to trigger the moment the cursor moves off the button, my naive solution would be to monitor for on_touch_down, and then monitor for on_touch_move (generally) that moves off the button (as well as on_touch_up inside the button to cancel).

If you want it to trigger the moment the touch is lifted outside the button, my naive solution would be to monitor for on_touch_down, and then monitor for on_touch_up (generally) and check if it is still in the button.

**Reply by RafaRioFal4699** [reply]
`The value of sender is unrelated to the message. It is always the button instance.`
Yes, that's my fault, and is totally removable. But that still doesn't explain why the event fires when I start, make and end the drag outside the button. Yes, the function is from the class, but event is from the button. Why the layout is getting an event assigned to its child? It's a propagation thing?

And thanks, for the patience, finally I could get my goal savingi the position where drag starts with `on_touch_down`, and validate if it's inside the button or not, letting the event trigger or not.

**Reply by Julian-O** [reply]
> Why the layout is getting an event assigned to its child? It's a propagation thing?

This is where I have to bow out, because you have reached the edge of my understanding; I am not an expert in event propagation in Kivy. I think I learnt more by [reading this a couple of days ago](https://github.com/orgs/kivy/discussions/14) than I knew before.

**Reply by RafaRioFal4699** [reply]
First of all, thank you. I have been following this framework for two years and thanks to you it is more alive than ever.

Well, having read the discussion you linked and having looked for event management in Android, I only have three options: 
 - I haven't implemented the events as I should (I'll work on it)
 - Assuming Kivy is doing it well, there is missing some documentation about event capturing and bubbling
 - There's a strange behavior that gives a root class the power of launch all of its children's events (and, relying on the explanation of the link, it should be corrected)
